deploy/cephfs/helm/templates/provisioner-clusterrole.yaml

```diff
@@ -25,10 +25,13 @@ rules:
   - apiGroups: [""]
     resources: ["events"]
     verbs: ["list", "watch", "create", "update", "patch"]
-  - apiGroups: [""]
-    resources: ["configmaps"]
-    verbs: ["get", "list", "create", "delete"]
   - apiGroups: ["csi.storage.k8s.io"]
     resources: ["csinodeinfos"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: [""]
```

Comment:
One of the things I kind of liked about having the attacher stuff totally separate was that it is easy to review and ensure some day when csidriver support is added we can easily cut out the attacher stuff entirely. This is starting to mix the use cases and will make it harder to do that.

If we're going to go down that path, can you please add a conditional around the changes to objects so that there is an {{ if .Values.attacher.enabled }} or something around it? We can easily find it then, and remove it later. Or, for those running it now, we could make {{ if not .Values.attacher.enabled }} create a CSIDriver object.

I'm not so sure what to do with the raw k8s resources though. maybe rootfs can weigh in?

Reply:
The problem I see (I could be wrong) with this approach of separating the Stateful sets is that, each stateful set can hence can run on different nodes, and so how do they exactly share the csi-provisioner.sock file, if at all.

I do understand that we are removing the ControllerPublish verbs, and hence there should be no request sent on the socket, but I believe there would at least be a GetPluginInfo call from the [attacher](https://github.com/kubernetes-csi/external-attacher/blob/927d0e3ee6a1089983c524b6c97ded69fe8ce51b/cmd/csi-attacher/main.go#L117), for which it may not receive any response if running on a node different than the provisioner, as there are no listeners on the mentioned socket.

Or is my reasoning incorrect?

Reply:
@ShyamsundarR in attacher statefulset we need to have 2 containers one for csi-attacher and another for ceph-plugin.

Reply:
IMO, This is breaking the assumptions around the various deployment models that CSI spec has detailed, or goes beyond stated deployment models of split and unified plugins.

In the suggestion [above](https://github.com/ceph/ceph-csi/pull/290#discussion_r270000587), I assume the controller is split into 2 parts, one a dummy for all purposes that responds to the attacher, stating that it does not support PUBLISH_UNPUBLISH_VOLUME, and the other that actually is the controller for create/delete/snapshot and other operations supported. In other words, we have a split the controller further into 2 parts (although the code and binary running within it are the same), one for the attacher and the other for the rest of the functionality.

I would think the right way to do this is as reflected in this PR at present, with enough conditionals in the Helm parts as suggested by @kfox1111 to make it easier to remove in the future, looks like the right way forward.

Reply:
> IMO, This is breaking the assumptions around the various deployment models that CSI spec has detailed, or goes beyond stated deployment models of split and unified plugins.
> 

i didn't get this one, how it will break the deployment models?

> In the suggestion [above](https://github.com/ceph/ceph-csi/pull/290#discussion_r270000587), I assume the controller is split into 2 parts, one a dummy for all purposes that responds to the attacher, stating that it does not support PUBLISH_UNPUBLISH_VOLUME, and the other that actually is the controller for create/delete/snapshot and other operations supported. In other words, we have a split the controller further into 2 parts (although the code and binary running within it are the same), one for the attacher and the other for the rest of the functionality.
> 
> I would think the right way to do this is as reflected in this PR at present, with enough conditionals in the Helm parts as suggested by @kfox1111 to make it easier to remove in the future, looks like the right way forward.

@ShyamsundarR adding a check in the helm will be fine, what we are going to do in Kube raw templates?

Reply:
> > IMO, This is breaking the assumptions around the various deployment models that CSI spec has detailed, or goes beyond stated deployment models of split and unified plugins.
> 
> i didn't get this one, how it will break the deployment models?

Exchanged a conversation on the csi slack channel on this topic. The "break" or differ from deployment models is that we will have our controller in 2 parts, which is not a documented deployment model in the spec.

> 
> > In the suggestion [above](https://github.com/ceph/ceph-csi/pull/290#discussion_r270000587), I assume the controller is split into 2 parts, one a dummy for all purposes that responds to the attacher, stating that it does not support PUBLISH_UNPUBLISH_VOLUME, and the other that actually is the controller for create/delete/snapshot and other operations supported. In other words, we have a split the controller further into 2 parts (although the code and binary running within it are the same), one for the attacher and the other for the rest of the functionality.
> > I would think the right way to do this is as reflected in this PR at present, with enough conditionals in the Helm parts as suggested by @kfox1111 to make it easier to remove in the future, looks like the right way forward.
> 
> @ShyamsundarR adding a check in the helm will be fine, what we are going to do in Kube raw templates?

kubernetes raw [templates](https://github.com/ceph/ceph-csi/tree/csi-v1.0/deploy/rbd/kubernetes) would have to be separated as per kubernetes versions, and the template will differ from the ones in Helm in case we want to add the conditionals within Helm.

Reply:
> the main intention I have added attacher as a sidecar container inside the provisioner statefulset was to avoid resource consumption (avoid deployment of extra statefulset)

Is an extra statefulset really that much overhead? I'm unconvinced.

It only needs to be a thing for 1.13.0 clusters too. Adding a csidriver object, we can avoid it completely on 1.14+. So I almost see it as a legacy thing we don't touch and it solves itself once we stop supporting 1.13?


> if needed i can deploy attacher as a separate statefulset with ceph-csi plugin as a container inside it, this will make us to keep both helm and raw templates almost similar, @rootfs @kfox1111 let me know your thoughts on that

I proposed a while ago the option of generating the static manifets using 'helm template' on the chart. That would entirely remove redundancy. I still think that is probably a good idea. I don't think anyone has had time to figure out the pluming to get it done though.
 
> can we maintain a separate directory structure for deployments to support different kubernetes version?

The template idea from above would make that much easier to maintain. Its a shame kustomize didn't hit until 1.14 or that could have been a solution to it too... Though maybe there is a way to use it that would work. they have includes I think. I haven't played with it much yet.

Reply:
> @ShyamsundarR in attacher statefulset we need to have 2 containers one for csi-attacher and another for ceph-plugin.

No, I don't think that is the case. See the https://github.com/kubernetes-csi/external-attacher --dummy option
They needed a way to skip attach when the driver didn't support controllerServer functionality. Before they had the flag attacher was still required by k8s, but has the dummy flag to make it not have to talk to the driver. Now that there is a CSIDriver the flag is there to skip needing the dummy attach pluming.

I don't think anyone's implemented that in ceph-csi yet. This PR would be a great place.

Reply:
> > @ShyamsundarR in attacher statefulset we need to have 2 containers one for csi-attacher and another for ceph-plugin.
> 
> No, I don't think that is the case. See the https://github.com/kubernetes-csi/external-attacher --dummy option
> They needed a way to skip attach when the driver didn't support controllerServer functionality. Before they had the flag attacher was still required by k8s, but has the dummy flag to make it not have to talk to the driver. Now that there is a CSIDriver the flag is there to skip needing the dummy attach pluming.
> 
> I don't think anyone's implemented that in ceph-csi yet. This PR would be a great place.

@kfox1111 i tried deploying csi-attcher with `--dummy` option as set  to true,  with this volume attachment to the pod still fails.
 my observation is that it expects the driver name should be equal to `csi/dummy`. 
```
	dummy             = flag.Bool("dummy", false, "Run in dummy mode, i.e. not connecting to CSI driver and marking everything as attached. Expected CSI driver name is \"csi/dummy\".")
```

as our driver name is `cephfs.ceph.com` it's skipping the volume attachment here https://github.com/kubernetes-csi/external-attacher/blame/master/pkg/controller/controller.go#L189